modules/compute/gke-node-pool/main.tf

```diff
@@ -253,3 +253,98 @@ resource "google_project_iam_member" "node_service_account_artifact_registry" {
   role    = "roles/artifactregistry.reader"
   member  = "serviceAccount:${local.sa_email}"
 }
+
+# Enable GPUDirect for A3 and A3Mega VMs, this involve multiple kubectl steps to integrate with the created cluster
+# 1. Install NCCL plugin daemonset
+# 2. Install NRI plugin daemonset
+# 3. Update user workload to inject rxdm sidecar and other required annotation, volume etc.
+locals {
+  split_cluster_id         = split("/", var.cluster_id)
+  user_workload_path_tcpx  = var.user_workload_path == null ? "${path.module}/gpu-direct-workload/sample-tcpx-workload-job.yaml" : var.user_workload_path
+  user_workload_path_tcpxo = var.user_workload_path == null ? "${path.module}/gpu-direct-workload/sample-tcpxo-workload-job.yaml" : var.user_workload_path
+
+  # Manifest to be installed for enabling TCPX on A3 machines
+  tcpx_manifests = [
+    "https://raw.githubusercontent.com/GoogleCloudPlatform/container-engine-accelerators/fee883360a660f71ba07478db95d5c1325322f77/gpudirect-tcpx/nccl-tcpx-installer.yaml",      # nccl_plugin v3.1.9 for tcpx
+    "https://raw.githubusercontent.com/GoogleCloudPlatform/container-engine-accelerators/fee883360a660f71ba07478db95d5c1325322f77/gpudirect-tcpx/nccl-config.yaml",              # nccl_configmap
+    "https://raw.githubusercontent.com/GoogleCloudPlatform/container-engine-accelerators/fee883360a660f71ba07478db95d5c1325322f77/nri_device_injector/nri-device-injector.yaml", # nri_plugin
+  ]
+
+  # Manifest to be installed for enabling TCPXO on A3Mega machines
+  tcpxo_manifests = [
+    "https://raw.githubusercontent.com/GoogleCloudPlatform/container-engine-accelerators/fee883360a660f71ba07478db95d5c1325322f77/gpudirect-tcpxo/nccl-tcpxo-installer.yaml",    # nccl_plugin v1.0.4 for tcpxo
+    "https://raw.githubusercontent.com/GoogleCloudPlatform/container-engine-accelerators/fee883360a660f71ba07478db95d5c1325322f77/nri_device_injector/nri-device-injector.yaml", # nri_plugin
+  ]
+
+  updated_user_workload_path = (
+    var.machine_type == "a3-highgpu-8g"
+    ? replace(local.user_workload_path_tcpx, ".yaml", "-tcpx.yaml")
+    : var.machine_type == "a3-megagpu-8g"
+    ? replace(local.user_workload_path_tcpxo, ".yaml", "-tcpxo.yaml")
+  : null)
+
+  gpu_direct_manifest = (
+    var.machine_type == "a3-highgpu-8g"
+    ? local.tcpx_manifests
+    : var.machine_type == "a3-megagpu-8g"
+    ? local.tcpxo_manifests
+  : null)
+}
+
+data "google_container_cluster" "gke_cluster" {
+  project  = var.project_id
+  name     = local.split_cluster_id[5]
+  location = local.split_cluster_id[3]
+}
+
+resource "null_resource" "install_dependencies" {
+  provisioner "local-exec" {
+    command = "pip3 install pyyaml argparse"
+  }
+}
+
+# execute script to inject rxdm sidecar into workload to enable tcpx for A3 VM workload
+resource "null_resource" "enable_tcpx_in_workload" {
```

Comment:
- I'd suggest to move all logic related to workloads outside the node-pool logic and only keep installation/setting up of anything related to node-pool itself. Maybe gke-job-template?

- You can then output the machinetype in gke-nodepool, and `use[]` it in the job-template module to decide if you need to manipulate the workload or not.

Reply:
we discussed this in the previous meeting. For now we will keep it in the same module, and we can move it out later if there is a need (like if there is more user provided parameters needed).